contracts/SP3QSAJQ4EA8WXEDSRRKMZZ29NH91VZ6C5X88FGZQ.crashpunks-v1.clar

```diff
@@ -0,0 +1,1004 @@
+        (if (< (len metaDataUrl) u10) (ok (var-get mint-counter))
+            (let
+                (
+                    ;; if client bypasses UI clientMintPrice then charge mint-price
+                    (myMintPrice (max-of (var-get mint-price) clientMintPrice))
+                    (mintCounter (var-get mint-counter))
+                    (index (get nft-index (map-get? nft-lookup {asset-hash: asset-hash, edition: u1})))
+                    (block-time (unwrap! (get-block-info? time u0) amount-not-set))
+                )
+                (asserts! (> maxEditions u0) editions-error)
+                (asserts! (> (stx-get-balance tx-sender) (var-get mint-price)) cant-pay-mint-price)
+                (asserts! (is-none index) asset-not-registered)
+
+                ;; Note: series original is really for later editions to refer back to this one - this one IS the series original
+                (map-insert nft-data {nft-index: mintCounter} {asset-hash: asset-hash, meta-data-url: metaDataUrl, max-editions: maxEditions, edition: u1, edition-cost: editionCost, series-original: mintCounter})
+
+                ;; Note editions are 1 based and <= maxEditions - the one minted here is #1
+                (map-insert nft-edition-counter {nft-index: mintCounter} {edition-counter: u2})
+
+                (if (> buyNowPrice u0)
+                    (map-insert nft-sale-data { nft-index: mintCounter } { sale-cycle-index: u1, sale-type: u1, increment-stx: u0, reserve-stx: u0, amount-stx: buyNowPrice, bidding-end-time: (+ block-height u200)})
+                    (map-insert nft-sale-data { nft-index: mintCounter } { sale-cycle-index: u1, sale-type: u0, increment-stx: u0, reserve-stx: u0, amount-stx: u0, bidding-end-time: u0})
+                )
+
+                (map-insert nft-lookup {asset-hash: asset-hash, edition: u1} {nft-index: mintCounter})
+
+                ;; The payment is split between the nft-beneficiaries with share > 0 they are set per edition
+                (map-insert nft-beneficiaries {nft-index: mintCounter} {addresses: addresses, shares: shares, secondaries: secondaries})
+
+                ;; finally - mint the NFT and step the counter
+                (if (is-eq tx-sender (var-get administrator))
+                    u0
+                    (unwrap! (paymint-split mintCounter myMintPrice tx-sender mintAddresses mintShares) payment-error-collection)
+                )
+                (unwrap! (nft-mint? crashpunks mintCounter tx-sender) failed-to-mint-err)
+                (print {evt: "mint-token", nftIndex: mintCounter, owner: tx-sender, amount: myMintPrice})
+                (var-set mint-counter (+ mintCounter u1))
+                (ok mintCounter)
+            )
+        )
+    )
+)
+
+(define-public (collection-mint-token (signature (buff 65)) (message-hash (buff 32)) (asset-hash (buff 32)) (metaDataUrl (string-ascii 256)) (maxEditions uint) (editionCost uint) (clientMintPrice uint) (buyNowPrice uint))
```

Comment:
`metaDataUrl` can be set by users freely. That value is returned by `get-token-uri`